nikola/shortcodes.py

```diff
@@ -209,14 +209,58 @@ def _parse_shortcode_args(data, start, shortcode_name, start_pos):
     raise ParsingError("Shortcode '{0}' starting at {1} is not terminated correctly with '%}}}}'!".format(shortcode_name, _format_position(data, start_pos)))
 
 
+def _extract_shortcodes(data):
+    """
+    Returns data, shortcodes:
+    
+    data is the original data, with the shortcodes replaced by UUIDs.
+
+    a dictionary of shortcodes, where the keys are UUIDs and the values
+    are the shortcodes themselves ready to process.
+    """
+    splitted = _split_shortcodes(data)
+    new_data = []
+    in_sc = False
+    buffer = []
+    shortcodes = {}
+    for chunk in splitted:
+        if chunk[0] == 'TEXT':
+            if in_sc:
+                buffer.append(chunk)
+            else:
+                new_data.append(chunk[1])
+        elif chunk[0] == 'SHORTCODE_START':
+            if in_sc:  # Previous sc doesn't close
+                sc_id = str(uuid.uuid4())
+                new_data.append(sc_id)
+                shortcodes[sc_id] = buffer[0][1]
+                for c in buffer[1:]:  # Dump buffered text
+                    new_data.append(c[1])
+            # Need to buffer to know if this SC closes
+            buffer = [chunk]
+            in_sc = True
+        elif chunk[0] == 'SHORTCODE_END':
+            # Previous shortcode closes here
+            buffer.append(chunk)
+            sc_id = str(uuid.uuid4())
+            new_data.append(sc_id)
```

Comment:
The UUID should be unique enough, considering there are more UUIDs than atoms in the universe. The stuff about dashes sounds good.